### PR TITLE
BIP 348: OP_CHECKSIGFROMSTACK

### PR DESCRIPTION
**Pull request by reardencode**
This BIP is based on the BCH OP_CHECKDATASIG work, as well as postings from the bitcoin dev mailing list in [this thread](https://www.mail-archive.com/bitcoin-dev@lists.linuxfoundation.org/msg10211.html). Some differences appear due to the activation of bips 340, 341, and 342 (taproot) since those were developed.

OP_CHECKSIGFROMSTACK is an OP_SUCCESS upgrade available only in BIP342 tapscript.

### CODE REVIEW

**Review by ElGhaly35 — COMMENTED** [review]
[``](`url`)

File: bip-csfs.mediawiki
Comment (ajtowns) [inline bip-csfs.mediawiki:10]: > Why is Public Domain no longer acceptable for new BIPs?
> 
>  * In some jurisdictions, public domain is not recognised as a legitimate legal action, leaving the BIP simply copyrighted with no redistribution or modification allowed at all.

File: bip-csfs.mediawiki
Comment (Roasbeef) [inline bip-csfs.mediawiki:21]: Why split across two extension mechanisms instead of just focusing on the modern path (op success)?

File: bip-csfs.mediawiki
Comment (Roasbeef) [inline bip-csfs.mediawiki:40]: Is the ref to BIP 137 here a typo? 137 is primarily about the old Bitcoin signed message scheme.

File: bip-csfs.mediawiki
Comment (Roasbeef) [inline bip-csfs.mediawiki:42]: Public key types were only introduced in segwit v1 ([BIP 342](https://github.com/bitcoin/bips/blob/master/bip-0342.mediawiki#cite_note-6)), so this wouldn't apply to a version of this op code that also exists under the segwit v0 paradigm.

File: bip-csfs.mediawiki
Comment (Roasbeef) [inline bip-csfs.mediawiki:49]: The sig ops budget is described in BIP 342: https://github.com/bitcoin/bips/blob/master/bip-0342.mediawiki#resource-limits

File: bip-csfs.mediawiki
Comment (Roasbeef) [inline bip-csfs.mediawiki:51]: I wonder if there's a use for a `OP_CHECKSIGFROMSTACKADD` variant...

So something that would be compatible with usage of `OP_CHECKSIGADD`. On the other hand, it is possible to replicate such behavior with `OP_CHECKSIGFROMSTACK` in combination with other existing op codes, eg: 
```
OP_TOALTSTACK (assuming the number N is on the top of the stack)
<pubkey> OP_CHECKSIGFROMSTACK
OP_IF
    OP_DROP (drop the 0x01 pushed in the success case)
    OP_FROMALTSACK (bring back the accumulator) 
    1 OP_ADD (acc += 1)
OP_ELSE
    OP_DROP (drop the empty vector) 
    OP_FROMALTSACK (put unchanged accumulator back on the stack)
OP_ENDIF
```

File: bip-csfs.mediawiki
Comment (Roasbeef) [inline bip-csfs.mediawiki:105]: Perhaps this should also credit Jeremy (as he made a mailing list post soliciting feedback for a design in 2021), and also the authors of the proposal he directly references in the ML?

File: bip-csfs.mediawiki
Comment (reardencode) [reply bip-csfs.mediawiki]: Signature verification for CHECKSIGFROMSTACK is done the same regardless of the script context.

File: bip-csfs.mediawiki
Comment (reardencode) [reply bip-csfs.mediawiki]: These legacy keys are no longer supported - sorry for the waste of time, this PR had gotten out of date while I was working on this document as [a BIN](https://github.com/bitcoin-inquisition/binana/blob/master/2024/BIN-2024-0003.md)

File: bip-csfs.mediawiki
Comment (reardencode) [reply bip-csfs.mediawiki]: Checking signatures from the stack is something that _can_ be supported on all script types (i.e. is compatible with NOP upgrade), and there are use cases where segwit v0 scripts still make the most sense vs. tapscripts.

If supporting legacy and segwitv0 where possible for new opcodes is not something that others value, switching CSFSV to an OP_SUCCESS upgrade would make sense.

File: bip-csfs.mediawiki
Comment (reardencode) [reply bip-csfs.mediawiki]: Yep. I had the same internal conversation and concluded that the rare cases where CSFSA behavior is needed using the altstack is an acceptable alternative.

Compared to standard CHECKSIG where CHECKMULTISIG was in active use on legacy/segwitv0 wallets, new wallets or protocols built on CSFS are much more likely to use MuSig2 or FROST when multiple signers are insolved, or (as BitGo's 2-of-3 MuSig2/tapscript wallet does) to have the different signing combinations split between tapscript paths where CSFSV+CSFS are more efficient than CSFSA.

**Comment by reardencode** [comment]
Updated to match the BIN, and with @Roasbeef 's comments. TYVM!

File: bip-csfs.md
Comment (murchandamus) [inline bip-csfs.md:9]: Please use the [preamble format](https://github.com/bitcoin/bips/blob/master/bip-0002.mediawiki#bip-header-preamble) specified in BIP2.

**Comment by reardencode** [comment]
Fixed preamble.

There's still 2 open conversations around this one:
* Should it add CSFSA (after discussions in ATX with Rusty and Rob, I'm inclined to add it)
* Should CSFSV be tapscript only (I'm inclined to add this functionality to legacy scripts as well, but may be in the minority there)

I'll post on the mailing list for additional opinions on these topics, but not sure whether it makes sense for the BIP to get numbered and merged as is while still potentially subject to those revisions.

**Comment by reardencode** [comment]
Fixed preamble.

There's still 2 open conversations around this one:
* Should it add CSFSA (after discussions in ATX with Rusty and Rob, I'm inclined to add it)
* Should CSFSV be tapscript only (I'm inclined to add this functionality to legacy scripts as well, but may be in the minority there)

I'll post on the mailing list for additional opinions on these topics, but not sure whether it makes sense for the BIP to get numbered and merged as is while still potentially subject to those revisions.

**Comment by murchandamus** [comment]
@reardencode: If you feel that your proposal is not ready to be merged, please convert the pull request to a Draft.

File: bip-csfs.md
Comment (ProofOfKeags) [inline bip-csfs.md:35]: nit: I'd like to offer the suggestion that the push order goes:

`<msg> <sig> <key>` instead of `<sig> <msg> <key>`. The current `CHECKSIG` ordering is `<sig> <key>` and it seems odd to me that we would make the CSFS opcode put the msg *between* these other pushes. Alternatively you could go `<sig> <key> <msg>` and it would seem to me as equally intuitive. As is though it feels like a gotcha.

Is there a specific reason this choice was made that I hadn't considered?

Comment (reardencode) [reply bip-csfs.md:35]: Anything that comes from the input stack should be before things that are provided by the locking script, so `sig` definitely comes first. Depending on particulars `msg` might come from the input stack or the locking script, and `key` always comes from the locking script. Examples:
* `<delegate-sig> <sig> <delegate-key> || DUP TOALT <key> CSFS FROMALT CHECKSIG`
* `<sig> <hash> || CTV <key> CSFS`
* `<sig> || <flags> TXHASH <key> CSFS`
* `<sig> <settlement-hash> <update-hash> || CTV SHA256 CAT <key> CSFS`

Comment (ProofOfKeags) [reply bip-csfs.md:35]: Retracted.

I understand the general direction of what you're getting at but I have to admit that I'm not following the specific use cases where `<msg>` might come from the witness instead of the script. It makes sense to me why you'd want `<msg>` in the script, what's lest clear is under what conditions it is useful to let the witness define it.

Comment (reardencode) [reply bip-csfs.md:35]: In any kind of delegation (see the first example above) the message is provided by the witness (i.e. the thing you're delegating to by signing). Here's a powerful example:

`<delegate_sig> <sig> <delegate_pubkey> <locktime> || CLTV SHA256 CAT <pubkey> CSFS VERIFY CHECKSIG`

Delegate to a key but only after `<locktime>`, with different locktimes per delegate key.

Comment (ProofOfKeags) [reply bip-csfs.md:35]: Forgive me but I have to expand this out because it doesn't make sense to me:

Witness:
1. delegate signature is pushed onto the stack: `<delegate_sig>`
2. the signature authorizing the delegate is pushed: `<delegate_sig> <authorizing_sig>`
3. the delegate pubkey is pushed: `<delegate_sig> <authorizing_sig> <delegate_pubkey>`
4. locktime is pushed: `<delegate_sig> <authorizing_sig> <delegate_pubkey> <locktime>`

Script:
5. locktime is checked: `<delegate_sig> <authorizing_sig> <delegate_pubkey> <locktime>`
6. locktime is hashed: `<delegate_sig> <authorizing_sig> <delegate_pubkey> <locktime_hash>`
7. locktime hash combined with delegate pubkey: `<delegate_sig> <authorizing_sig> <delegate_pubkey || locktime_hash>`
8. authorizing pubkey is pushed: `<delegate_sig> <authorizing_sig> <delegate_pubkey || locktime_hash> <authorizing_pubkey>`
9. CSFS(V) operation(s) checks that the delegate pubkey is authorized after the locktime by checking the `key || locktime` message was signed by `authorizing_sig`: `<delegate_sig>`
10. CHECKSIG operation fails because we don't have a key? Maybe you forgot a dup somewhere? Maybe I don't understand script as well as I thought? idk.

Assuming this is roughly what you meant, then I'd imagine in this case the authorizing signature would be pre-signed and distributed. Then at transaction time the delegate presents the authorizing signature along with their key and assigned locktime.

Assuming all of this is a fair reading, then I think I get what you mean with `<msg>` having an interesting use case wherein it could be:
1. provided by the script
2. provided by the witness
3. collaboratively derived from witness and script participation.

... which seems pretty cool. I'd love if you can correct either my understanding or your example for completeness, but otherwise I'm satisfied with this and this thread can be resolved.

**Review by Jesus5432 — APPROVED** [review]

**Comment by murchandamus** [comment]
I converted this to draft since @reardencode indicated that this would likely need more changes before being ready for number assignment and merge. Given that this hasn’t been updated in half a year, is this still in progress?

**Comment by reardencode** [comment]
Definitely still WIP. Just hasn't been much reason to advance the conversation around the two specific questions I mentioned given the state of the broader conversation around bitcoin's future consensus changes.

I am a bit confused about the need to have those questions answered before merging. Is the intention of the BIPs repository only to hold final specifications for consensus changes, or to enable iterating on consensus changes before proposing a document for numbering and merging?

**Comment by murchandamus** [comment]
> I am a bit confused about the need to have those questions answered before merging. Is the intention of the BIPs repository only to hold final specifications for consensus changes, or to enable iterating on consensus changes before proposing a document for numbering and merging?


You had written above

> I'll post on the mailing list for additional opinions on these topics, but not sure whether it makes sense for the BIP to get numbered and merged as is while still potentially subject to those revisions.

which I interpreted as a request to hold off while the situation might still be developing. I take it that I misunderstood you?

**Comment by reardencode** [comment]
> which I interpreted as a request to hold off while the situation might still be developing. I take it that I misunderstood you?

Gotcha. Yeah, I honestly just don't know whether it makes sense to number and merge the BIP with outstanding questions like those or not. Since we're back in motion LNHANCE, I'll follow through and see what opinions appear :)

Thanks!

**Comment by murchandamus** [comment]
@reardencode: Preferably, as much as possible of the proposal is in place before it is put forth for an Editor review. Open questions tend to signal that a BIP may still be developing directionally and editors may be holding off due to that. ;)

If the open questions concern minor details, or have been resolved, please let us know that it is ready for an editor review.

**Comment by moonsettler** [comment]
> Fixed preamble.
> 
> There's still 2 open conversations around this one:
> 
>     * Should it add CSFSA (after discussions in ATX with Rusty and Rob, I'm inclined to add it)
> 
>     * Should CSFSV be tapscript only (I'm inclined to add this functionality to legacy scripts as well, but may be in the minority there)
> 
> 
> I'll post on the mailing list for additional opinions on these topics, but not sure whether it makes sense for the BIP to get numbered and merged as is while still potentially subject to those revisions.

For what it's worth, I also think signature aggregation will be the dominant
form of CSFS use. LNhance at it's core is CTV + CSFS, and so it makes sense
to have both of those available in pre-tapscript.

No strong opinion on CHECKSIGFROMSTACKADD, agree with the general reasoning.

It's a bit weird to backport Schnorr this way, and the NOP upgrade path
leaving 3 elements on the stack is also unfortunate. On the other hand,
reverting CSFSV to use ECDSA in pre-tapscript would force us to consider
implementing script multisig, to do anything really worthwhile there.

File: bip-csfs.md
Comment (reardencode) [reply bip-csfs.md:35]: Sorry for the long delay. You're quite correct that the example I gave does not work. Apparently I should have waited for Taproot Wizards new neovim script plugin. I believe this works for 3 more opcodes :(

`<sig> <delegate_sig> <delegate_pubkey> <locktime> || CLTV SHA256 OVER CAT TOALT CHECKSIGVERIFY FROMALT <pubkey> CSFS`

1. the signature authorizing the delegate is pushed: `<authorizing_sig>`
2. delegate signature is pushed onto the stack: `<authorizing_sig> <delegate_sig>`
3. the delegate pubkey is pushed: `<authorizing_sig> <delegate_sig> <delegate_pubkey>`
4. locktime is pushed: `<authorizing_sig> <delegate_sig> <delegate_pubkey> <locktime>`
5. locktime is checked: `<authorizing_sig> <delegate_sig> <delegate_pubkey> <locktime>`
6. locktime is hashed: `<authorizing_sig> <delegate_sig> <delegate_pubkey> <locktime_hash>`
7. delegate pubkey is copied over: `<authorizing_sig> <delegate_sig> <delegate_pubkey> <locktime_hash> <delegate_pubkey>`
8. locktime hash combined with delegate pubkey: `<authorizing_sig> <delegate_sig> <delegate_pubkey> <locktime_hash || delegate_pubkey>`
9. concatenation is moved to alt stack: `<authorizing_sig> <delegate_sig> <delegate_pubkey>`
10. CHECKSIGVERIFY validates the signature from the delegated key: `<authorizing_sig>`
11. concatenation is moved from alt stack: `<authorizing_sig> <locktime_hash || delegate_pubkey>`
12. authorizing pubkey is pushed: `<authorizing_sig> <locktime_hash || delegate_pubkey> <authorizing_pubkey>`
10. CSFS operation(s) checks that the delegate pubkey is authorized after the locktime by checking the `locktime || key` message was signed by `authorizing_sig`: `TRUE`

Thanks for checking the script.

**Comment by moonsettler** [comment]
Can we change the name of this PR and remove (VERIFY) to more accurately reflect the latest state of the proposal?

File: bip-csfs.mediawiki
Comment (reardencode) [reply bip-csfs.mediawiki]: Keeping this BIP to 1 opcode. Other CSFS(A/V) variants can be added later if they are specifically needed.

**Comment by reardencode** [comment]
Updated with an additional example, removed CSFSV, and all earlier review comments addressed.

File: bip-csfs.md
Comment (murchandamus) [inline bip-csfs.md:24]: This mentions OP_CSFSV, was that supposed to say just `OP_CHECKSIGFROMSTACK` or is this sentence obsolete?

File: bip-csfs.md
Comment (murchandamus) [inline bip-csfs.md:49]: Is this "(VERIFY)" obsolete or shorthand for `OP_CSFS OP_VERIFY` here?

```suggestion
3. Add/multisig: No concession is made to `OP_CHECKMULTISIG` or `OP_CHECKSIGADD` semantics with `OP_CHECKSIGFROMSTACK`. In Tapscript, add semantics can be implemented with 1 additional vByte per key (`OP_TOALTSTACK OP_CHECKSIGFROMSTACK OP_FROMALTSTACK OP_ADD`).
```

File: bip-csfs.md
Comment (murchandamus) [inline bip-csfs.md:28]: This feels a bit out-of-context because keys have not been mentioned at all so far.

File: bip-csfs.md
Comment (murchandamus) [inline bip-csfs.md:38]: I’m not sure whether this is intentional, but previously we treat a 32-byte key as a BIP 340 public key. I assume if it were out-of-range, the signature validation would fail.
However here, rather than repeating that the lengths 0-byte and 32-byte are excluded, the criterion appears to be "not zero and not a BIP 340 public key". I assume this is covered by the previous line, but what if a 32-byte value were provided that is out of range, would this line apply or not? 

Is it possible that this should read:

```suggestion
* If the public key size is not zero bytes and not 32 bytes; the public key is of an unknown public key type. Signature verification for unknown public key types succeeds as if signature verification for a known public key type had succeeded.
```

File: bip-csfs.md
Comment (murchandamus) [inline bip-csfs.md:50]: It would be easier to follow if the title of BIP 347 were mentioned:

```suggestion
4. Splitting R/S on the stack: Implementing split/separate signatures is left as an exercise for other bitcoin upgrades, such as [BIP 347] (OP_CAT).
```

It also isn’t clear to me why this link text is code formatted.

**Review by murchandamus — COMMENTED** [review]
Let’s call this BIP 348.

File: bip-csfs.md
Comment (murchandamus) [reply bip-csfs.md]: Ah, my bad, it’s not that it’s not merged, but that it is a markdown file rather than mediawiki.

File: bip-0348.md
Comment (murchandamus) [inline bip-0348.md:127]:
```suggestion
[BIP 349]: https://github.com/bitcoin/bips/blob/master/bip-0349.md
```

File: bip-csfs.md
Comment (reardencode) [reply bip-csfs.md]: I believe I copied that text from BIP340. Your wording is more clear.

File: bip-csfs.md
Comment (reardencode) [reply bip-csfs.md]: am confused, the PR shows merged.

Comment (reardencode) [reply bip-csfs.md]: ah, had the link wrong - `md`

File: bip-csfs.md
Comment (reardencode) [reply bip-csfs.md]: For the purposes of Lightning Symmetry, when not attempting to build an alternative Lightning Symmetry that has endogenous fees, CSFS+CTV are a minimal substitute for APO. CSFS+CTV+PC+IKEY (LNHANCE) are byte-wise and composability-wise better substitute to APO.

Comment (reardencode) [reply bip-csfs.md]: Added a brief summary of this to the BIP

**Comment by murchandamus** [comment]
> Would love you to read over the new text I added in the Lightning Symmetry section.

LGTM

image

Nit: In the future, if you don’t reposition all the linebreaks, it’s much easier to see what was changed in text. :nerd_face:

I.e.

```diff
-`OP_CHECKSIGFROMSTACK` (CSFS) can be used in Lightning Symmetry channels.
-The construction `OP_CHECKTEMPLATEVERIFY <pubkey> OP_CHECKSIGFROMSTACK` with a
-spend stack containing the CTV hash and a signature for it is logically
-equivalent to `<bip118_pubkey> OP_CHECKSIG` and a signature over
-`SIGHASH_ALL|SIGHASH_ANYPREVOUTANYSCRIPT`. The `OP_CHECKSIGFROMSTACK`
-construction is 8 vBytes larger.
+`OP_CHECKSIGFROMSTACK` (CSFS) can be used to implement Lightning Symmetry
+channels. The construction `OP_CHECKTEMPLATEVERIFY <pubkey>
+OP_CHECKSIGFROMSTACK` with a spend stack containing the CTV hash and a
+signature for it is logically equivalent to `<bip118_pubkey> OP_CHECKSIG` and
+a signature over `SIGHASH_ALL|SIGHASH_ANYPREVOUTANYSCRIPT`. The
+`OP_CHECKSIGFROMSTACK` construction is 8 vBytes larger.
```

is harder to parse than

```diff
-`OP_CHECKSIGFROMSTACK` (CSFS) can be used in Lightning Symmetry channels.
+`OP_CHECKSIGFROMSTACK` (CSFS) can be used to implement Lightning Symmetry channels.
```

**Comment by reardencode** [comment]
@murchandamus I believe moon and a couple others will take a look today, then I think this is G2G.

**Review by moonsettler — APPROVED** [review]
ACK on the current content.

~Might want to consider mentioning the deleted key function verification scheme we learned from @bigspider?~

**edit:**
Will add to PAIRCOMMIT instead! Since it needs multiple elements signed to work.

**Review by Jesus5432 — APPROVED** [review]

**Review by Jesus5432 — APPROVED** [review]

**Pull request merged**
